core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -674,135 +675,346 @@ private RexNode simplifyCoalesce(RexCall call) {
+        if (!o.accept(this)) {
+          return false;
         }
       }
-      RexNode disjunction = RexUtil.composeDisjunction(rexBuilder, terms);
-      if (!call.getType().equals(disjunction.getType())) {
-        disjunction = rexBuilder.makeCast(call.getType(), disjunction);
+      return true;
+    }
+
+    @Override public Boolean visitOver(RexOver over) {
+      return false;
+    }
+
+    @Override public Boolean visitCorrelVariable(RexCorrelVariable correlVariable) {
+      return false;
+    }
+
+    @Override public Boolean visitDynamicParam(RexDynamicParam dynamicParam) {
+      return false;
+    }
+
+    @Override public Boolean visitRangeRef(RexRangeRef rangeRef) {
+      return false;
+    }
+
+    @Override public Boolean visitFieldAccess(RexFieldAccess fieldAccess) {
+      return true;
+    }
+
+    @Override public Boolean visitSubQuery(RexSubQuery subQuery) {
+      return false;
+    }
+
+    @Override public Boolean visitTableInputRef(RexTableInputRef fieldRef) {
+      return false;
+    }
+
+    @Override public Boolean visitPatternFieldRef(RexPatternFieldRef fieldRef) {
+      return false;
+    }
+
+  }
+
+  /** Analyzes a given rexnode - and decides whenever its casesafe to unwind.
+  *
+  * CaseSafe means that it only contains a combination of known good oeprators.
+  *
+  * The divison is an unsafe operator; consider: case when a>0 then 1/a else null end
+  */
+  static boolean isSafeExpression(RexNode r) {
+    return r.accept(new SafeRexVisitor());
+  }
+
+  private static RexNode simplifyBooleanCase(RexBuilder rexBuilder,
+      List<CaseBranch> inputBranches, RexUnknownAs unknownAs, RelDataType t) {
+    RexNode result = null;
+
+    // prepare all condition/branches for boolean interpretation
+    // It's done here make these interpretation changes available to case2or simplifications
+    // but not interfere with the normal simplifcation recursion
+    List<CaseBranch> branches = new ArrayList<>();
+    for (CaseBranch branch : inputBranches) {
+      if (!isSafeExpression(branch.cond) || !isSafeExpression(branch.value)) {
+        return null;
+      }
+      RexNode cond;
+      RexNode value;
+      if (branch.cond.getType().isNullable()) {
+        cond = rexBuilder.makeCall(SqlStdOperatorTable.IS_TRUE, branch.cond);
+      } else {
+        cond = branch.cond;
       }
-      return simplify(disjunction, unknownAs);
+      if (!t.equals(branch.value.getType())) {
+        value = rexBuilder.makeAbstractCast(t, branch.value);
+      } else {
+        value = branch.value;
+      }
+      branches.add(new CaseBranch(cond, value));
     }
-    if (newOperands.equals(operands)) {
-      return call;
+
+    // 1) Possible simplification if unknown is treated as false:
+    //   CASE
+    //   WHEN p1 THEN TRUE
+    //   WHEN p2 THEN TRUE
+    //   ELSE FALSE
+    //   END
+    // can be rewritten to: (p1 or p2)
+    if (unknownAs == FALSE) {
+      result = simplifyBooleanCase1(rexBuilder, branches);
```

Comment:
How `simplifyBooleanCase1` differs from `simplifyBooleanCase2` and from `simplifyBooleanCase3`?

It looks like `simplifyBooleanCase3` covers all the cases, so 1 and 2 variations can be just removed

Reply:
I was not looking to remove redundancy earlier; but yes there is: case2 handles all cases of case1 more generally.
case3 works a little bit differently; and although it simplifies case1/case2 variants; it leaves a bit more complicated expressions behind. We might be later able to get rid of case2 and keep only the most general - but that will need finer grained and/or simplification; which we might do later - but until then I would recommend to keep case2.

I've done the following changes:

* removed case1 entirely and extended case2's description to show that it also includes case1
* renamed case2 to "simplifyBooleanCaseBooleanBranches"
* renamed case3 to "simplifyBooleanCaseGeneric" - (I'm not a big fan of mile long function names; but in this case they are only used to arrange logic, let me know if you think the "numbered" was better)
* I've also moved the comments to be apidoc-s instead of just notes - I think since they have a dedicated function; it might be better.

Reply:
>case3 works a little bit differently; and although it simplifies case1/case2 variants; it leaves a bit more complicated expressions behind.

What's the point of having two different code branches that do the same thing?

>  it leaves a bit more complicated expressions behind

Is it caused by the fact that "case3" does not call `simplify` over the created expressions?
Why don't you do that?

Reply:
The key difference is that for the first one every branch is either true or false; consider:
```
case when c1 then true
     when c2 then true
     else false
end
```
This is rewritten to `(c1 or c2)`  by the boolean-branches one; but to `(c1 or (c2 and not(c1))` by the general.
In case `c1` is not just a simple expression which we could handle as predicate; we won't be able to reduce the latter back to `(c1 or c2)`.

Reply:
In general, one can collapse

```
case
  when c1 then X
  when c2 then X
...
end
```

into 

```
case
  when c1 or c2 then X
...
end
```

Reply:
In other words, consequent case branches with the same value can be collapsed provided relevant when expressions are safe.

This applies for all kinds of CASE expressions, not just for booleans.

Reply:
yes, it would make sense to compact subsequent branches with the same conclusion - this would be essentialy the generaliziation of the value cardinality check.

but there is one more case in which the general would perform a little bit worse:
```
case
  when c1 then true
  when c2 then false
  when c3 then true
  else false
end
```
would be `c1 or (c3 and not(c2))` by the booleanBranches and `c1 or (c2 and not(c2) and not(c1)` by the general - but I think I can fuse the 2 logics together...without breaking this case.